rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/delay/TbMsgDelayNodeTest.java

```diff
@@ -0,0 +1,332 @@
+    public void givenInvalidPeriodTimeUnits_whenInit_thenNodeFailsToStartWithCorrectException(TimeUnit periodTimeUnit) {
+        // GIVEN
+        config = config.defaultConfiguration();
+        config.setPeriodTimeUnit(periodTimeUnit);
+
+        // WHEN-THEN
+        assertThatThrownBy(() -> node.init(ctxMock, new TbNodeConfiguration(JacksonUtil.valueToTree(config))))
+                .isInstanceOf(TbNodeException.class)
+                .hasMessage("Unsupported time unit: " + periodTimeUnit)
+                .matches(e -> ((TbNodeException) e).isUnrecoverable());
+    }
+
+    @Test
+    public void givenTwoMsgsAndLimitOfOne_whenOnMsg_thenShouldDelayFirstAndFailSecond() {
+        // GIVEN
+        config = config.defaultConfiguration();
+        config.setMaxPendingMsgs(1);
+
+        var selfId = new RuleNodeId(UUID.randomUUID());
+
+        try {
+            node.init(ctxMock, new TbNodeConfiguration(JacksonUtil.valueToTree(config)));
+        } catch (TbNodeException e) {
+            fail("Node failed to start!", e);
+        }
+
+        given(ctxMock.getSelfId()).willReturn(selfId);
+
+        var expectedTickMsg = TbMsg.newMsg(
+                null, TbMsgType.DELAY_TIMEOUT_SELF_MSG, selfId,
+                msg.getCustomerId(), TbMsgMetaData.EMPTY, msg.getId().toString()
+        );
+        given(ctxMock.newMsg(
+                isNull(), eq(TbMsgType.DELAY_TIMEOUT_SELF_MSG), eq(selfId),
+                eq(msg.getCustomerId()), eq(TbMsgMetaData.EMPTY), eq(msg.getId().toString())
+        )).willReturn(expectedTickMsg);
+
+        // WHEN 1: first message is received
+        var firstMsg = msg;
+
+        node.onMsg(ctxMock, firstMsg);
+
+        // THEN 1: first message is correctly delayed
+        ArgumentCaptor<TbMsg> tickMsgCaptor = ArgumentCaptor.forClass(TbMsg.class);
+        then(ctxMock).should(times(1)).tellSelf(tickMsgCaptor.capture(), anyLong());
+        assertThat(tickMsgCaptor.getValue()).isEqualTo(expectedTickMsg);
+        then(ctxMock).should(times(1)).ack(firstMsg);
+
+        // WHEN 2: second message is received
+        var secondMsg = newMsg();
+
+        node.onMsg(ctxMock, secondMsg);
+
+        // THEN 2: seconds message fails because of the limit
+        var failedMsgCaptor = ArgumentCaptor.forClass(TbMsg.class);
+        var exceptionCaptor = ArgumentCaptor.forClass(Exception.class);
+        then(ctxMock).should(times(1)).tellFailure(failedMsgCaptor.capture(), exceptionCaptor.capture());
+        assertThat(failedMsgCaptor.getValue()).isEqualTo(secondMsg);
+        assertThat(exceptionCaptor.getValue()).isInstanceOf(RuntimeException.class);
+        assertThat(exceptionCaptor.getValue()).hasMessage("Max limit of pending messages reached!");
+
+        // WHEN 3: tick message from delayed first message is received
+        node.onMsg(ctxMock, expectedTickMsg);
+
+        try {
```

Comment:
Remove sleep. Please add a screenshot that shows that the tests passed 100 times.

Reply:
![image](https://github.com/thingsboard/thingsboard/assets/99619831/e48e509b-6a0c-4ecc-89e1-8c5a94206c79)